File: ml-engine/app/schemas/transaction.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class TransactionEvent(BaseModel):
    transactionId: str
    amount: float
    currency: Optional[str] = None
    sourceAccountId: Optional[str] = None
    destinationAccountId: Optional[str] = None
    channel: Optional[str] = None
    country: Optional[str] = None
    city: Optional[str] = None
    timestamp: Optional[datetime] = None
    customerId: Optional[str] = None
    sourceBalanceBefore: Optional[float] = None
    sourceBalanceAfter: Optional[float] = None
    destinationBalanceBefore: Optional[float] = None
    destinationBalanceAfter: Optional[float] = None

File: backend/transaction-service/src/main/java/com/frauddetection/transaction/entity/Transaction.java
package com.frauddetection.transaction.entity;

import com.frauddetection.common.enums.TransactionStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.Instant;

@Entity
@Table(name = "transactions")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Transaction {

    @Id
    @Column(name = "id", nullable = false, updatable = false)
    private String id;

    @Column(nullable = false)
    private BigDecimal amount;

    @Column(nullable = false, length = 3)
    private String currency;

    @Column(name = "source_account_id", nullable = false)
    private String sourceAccountId;

    @Column(name = "destination_account_id", nullable = false)
    private String destinationAccountId;

    @Column(nullable = false)
    private String channel;

    private String country;

    private String city;

    @Column(name = "customer_id")
    private String customerId;

    @Column(name = "source_balance_before")
    private BigDecimal sourceBalanceBefore;

    @Column(name = "source_balance_after")
    private BigDecimal sourceBalanceAfter;

    @Column(name = "destination_balance_before")
    private BigDecimal destinationBalanceBefore;

    @Column(name = "destination_balance_after")
    private BigDecimal destinationBalanceAfter;

    @Column(nullable = false)
    private Instant timestamp;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionStatus status;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;
}
